Source/WebCore/Modules/cookie-store/CookieStore.cpp

```diff
@@ -210,79 +241,111 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
         return;
     }
 
-    if (!m_cookieJar) {
-        promise->reject(SecurityError);
-        return;
-    }
-
-    auto& document = *downcast<Document>(context);
-    // The maximum attribute value size is specified at https://wicg.github.io/cookie-store/#cookie-maximum-attribute-value-size.
-    static constexpr auto maximumAttributeValueSize = 1024;
+    ALWAYS_LOG_WITH_STREAM(stream << "rupin log - point 1");
 
-    Cookie cookie;
-    cookie.name = WTFMove(options.name);
-    cookie.value = WTFMove(options.value);
-    cookie.created = WallTime::now().secondsSinceEpoch().milliseconds();
+    auto setCookies = [options = WTFMove(options), promise = WTFMove(promise)](ScriptExecutionContext& context) {
+        ALWAYS_LOG_WITH_STREAM(stream << "rupin log - point 3 - lambda called");
 
-    cookie.domain = options.domain.isNull() ? document.domain() : WTFMove(options.domain);
-    if (!cookie.domain.isNull()) {
-        if (cookie.domain.startsWith('.')) {
-            promise->reject(Exception { TypeError, "The domain must not begin with a '.'"_s });
+        auto& document = downcast<Document>(context);
+        auto* page = document.page();
+        if (!page) {
+            promise->reject(SecurityError);
             return;
         }
 
-        if (!m_host.endsWith(cookie.domain) || (m_host.length() > cookie.domain.length() && !StringView(m_host).substring(0, m_host.length() - cookie.domain.length()).endsWith('.'))) {
-            promise->reject(Exception { TypeError, "The domain must be a part of the current host"_s });
-            return;
+        // The maximum attribute value size is specified at https://wicg.github.io/cookie-store/#cookie-maximum-attribute-value-size.
+        static constexpr auto maximumAttributeValueSize = 1024;
+
+        Cookie cookie;
+        cookie.name = WTFMove(options.name);
+        cookie.value = WTFMove(options.value);
+        cookie.created = WallTime::now().secondsSinceEpoch().milliseconds();
+
+        cookie.domain = options.domain.isNull() ? document.domain() : WTFMove(options.domain);
+        if (!cookie.domain.isNull()) {
+            if (cookie.domain.startsWith('.')) {
+                promise->reject(Exception { TypeError, "The domain must not begin with a '.'"_s });
+                return;
+            }
+
+            auto host = document.url().host().toString();
+            if (!host.endsWith(cookie.domain) || (host.length() > cookie.domain.length() && !StringView(host).substring(0, host.length() - cookie.domain.length()).endsWith('.'))) {
+                promise->reject(Exception { TypeError, "The domain must be a part of the current host"_s });
```

Comment:
This is not OK in the worker case. You're rejecting the promise on the main thread instead of the worker thread.